### DISCUSSION TITLE
magit-blame-removal and magit-blame-reverse has some functionality swapped?

### DISCUSSION THREAD

**Discussion by geza-herman** [magit/magit#5017] [Ideas]
Note, I'm not sure whether this is a bug, but the current behavior seems odd to me. Or maybe some improvement on the documentation could help in this case, if the current behavior is intended.

So, what is the issue? Look at the 3 blaming kinds:
- *addition*: call `magit-blame-addition`. After, when calling `magit-show-commit` on a chunk, it shows the commit that caused the change. If `magit-blame-addition` is called again, then magit will jump to a previous rev which doesn't have the modifications of the commit (it jumps to the parent of the commit, supposedly). That's fine.
- *removal*: call `magit-blame-removal`. Here, when calling `magit-show-commit` on a chunk, it shows the commit of the next modification. But, if `magit-blame-removal` is called the 2nd time, it will stay on the current revision, it doesn't go forward. So *removal* is not the opposite of *addition*. In one regard, it is similar to *addition*, because `magit-show-commit` shows the commit that made the change; but in the other hand, it is not, because it doesn't jump to a state which has the change incorporated.
- *reverse*: call `magit-blame-reverse`. Here, when calling `magit-show-commit` on a chunk, it shows the preceding commit to the next modification (so, in a way, we can say that it's a totally unrelated commit). And when `magit-blame-reverse` is called, then magit will jump to the changing commit, which has the change of the commit incorporated.

So it seems that the rev that *removal* and *reverse* jump is the other way around than it should be.

When *addition* is used to go backwards in history, the information presented is in sync: `magit-blame-addition` jumps to a state, which differs by the diff shown by `magit-show-commit` (regarding the chunk). But neither *removal* nor *reverse* can be used to go forwards in history, while `magit-show-commit` showing the diff of the current and the next state (*removal* cannot go continuously, it stops after the first call. *reverse* can be used to walk forwards in history, but `magit-show-commit` shows a diff which has nothing to do with the diff of the current and next state).

**Comment by tarsius** [comment]
I don't use blame a lot anymore, at least not to move more than one step. That's because I agree with you that recursive blaming is not exactly intuitive.

I now use "tracing" instead. `magit-log-trace-definition` (`C-c f t` with the recommended bindings), shows a lot of all the commits that touched the current definition. This requires `diff.LANG.xfuncname` is set and that there is a mapping for `LANG` in a git "attributes" file. Alternatively you can use `magit-log-buffer-file` (`C-c f l`) with a region to show the history of that. That maybe sometimes be necessary even if tracing is configured; there might be multiple things with the same name (e.g., a variable and a function, or maybe multiple methods for the same generic function), and tracing always picks the first match for the regexp specified by `xfuncname`.

But back to blaming. (This is mostly just a collection of things that occurred to my while playing with blaming. In the process I have noticed that some things could be improved, in particular when switching from the file/blob to the diff, we don't always jump to the correct location. But I won't address every point you raised right now. I think there is room for improvement, but I will delay working on that until I have more time. Brain dump:)

I did some experiments and things work as intended and documented. The documentation could be more detailed, of course. E.g., it is not wrong to say that all the variants can be used recursively, even though you have experienced something else. It's just that you might have to move to a different location in the blamed file for that to make sense. (Or maybe it would always make sense even when staying on the same line. I'll have to think about that some more.) The documentation should also cover workflows, some aspects I am briefly touching on here.

If you have gone backward in history using the *addition* variant a few times, and simply want to go back in the opposite direction you came, use `q`.

The *removal* is internally also known as *final*.It shows the final commit that still had the commit. It is indeed highly likely that commit has nothing to do with the commit that *removes* the line in question, except that it is a parent of that commit. This is a direct mapping of `git blame --reverse`, using its output as intended.

The *reverse* variant is extremely similar to *removal*, infact the only difference is that if flips the two commits returned by `git blame --reverse` (implemented in `magit-blame--parse-chunk`). The result is that `magit-show-commit`, shows the commit that removes the line, not the first that lacks it.

You can toggle between the *removal* and *reverse* variants, to change what commit is shown when pressing `RET`. [Note to self: avoid running re-blaming; just flip the commits in all chunk objects and update display.]

I believe the *reverse* variant is also more useful for recursive blaming, just not without first moving to another line. So that's two improvements over `git blame`'s own back to the future time travel.

But I get a feeling that this could be further improved - along the same lines you seem to have in mind. I.e., a new variant could be added that really jumps to the first version that doesn't have the line anymore. An obvious issue that will need addressing that jumping to the line that no longer exists, is not really possible. We want to jump to the line that replaced it, but that might not be a 1:1 mapping.

The first step will be to fix the jumping to the same location when going to the diff, which I mentioned above. The second step will be jumping to the correct location when going forward as just mentioned. Like it is currently possible to toggle between the *removal*/*reverse* variants, it should be possible between *reverse* and the new (working title) *actually-gone* variant. This will be more complex, because we cannot just flip the commits, we have to switch between two blobs containing different blame information. So more "jumping to the 'same location' (which doesn't exist anymore") :stuck_out_tongue_closed_eyes:.

I don't plan to work on this soon. I also consider this a feature request, which is why I converted it to a discussion.

**Comment by geza-herman** [comment]
(sorry for the late response, I didn't notice your answer)

I also like the tracing feature, but I think it's not too reliable (it's a git problem, not a magit one). If the code got significant changes (for example, moving around things, while renaming them), then tracing becomes confused, and I got a trace which misses things. Recursive blaming also has this problem, but at least in this case I have more opportunities to dig in (I have more context because I see the whole file, I see all the other blame information in the file, I have more chance to jump to the correct place).

Regarding `q` to move in the opposite direction: thanks for the suggestion, it's an option. But it only works if I started blaming with a recent version. But if I started with an old revision, I do need to be able to move forward.

I think that you swapped *removal*/*reverse* at some places. At least, the *final* variant is the alias of *reverse*, and *reverse* is the variant which shows the last commit which still had the line in place.

> But I get a feeling that this could be further improved - along the same lines you seem to have in mind. I.e., a new variant could be added that really jumps to the first version that doesn't have the line anymore.

But magit already has this, doesn't it? I mean, if I use `magit-blame-reverse`, it does that. If I start blaming with `magit-blame-reverse`, it shows the commit before the one which removes the line. But, when `magit-blame-reverse` run again, it will jump to the commit which doesn't have the line anymore. That is my problem: magit shows a different commit in the blame buffer than it jumps to. So, I have to use `magit-blame-removal` to see which commit removed a line, and then I have to use `magit-blame-reverse` twice to actually jump to this commit (the first call switches to reverse mode, and the next one does the jump).

**Comment by tarsius** [comment]
Given that I believe the answer to

> magit-blame-removal and magit-blame-reverse has some functionality swapped?

is "no", I am closing this. (I've also added the label, which reminds me to have another look, should I find some extra time under some rock.)
